core/collection_mode.ml

```diff
@@ -1,28 +1,117 @@
 open! Core
 
+module Primary = struct
+  type t =
+    | Intel_processor_trace
+    | Stacktrace_sampling
+end
+
+module Event = struct
+  module Name = struct
+    type t =
+      | Branch_misses
+      | Cache_misses
+    [@@deriving compare, hash, sexp]
+
+    let to_string = function
+      | Branch_misses -> "branch-misses"
+      | Cache_misses -> "cache-misses"
+    ;;
+  end
+
+  module Config = struct
+    type t =
+      | Frequency of int
+      | Period of int
+    [@@deriving of_sexp]
+  end
+
+  type t =
+    { config : Config.t
+    ; name : Name.t
+    }
+  [@@deriving of_sexp]
+
+  let of_string = function
+    | "branch-misses" -> { config = Period 50; name = Branch_misses }
+    | "cache-misses" -> { config = Period 50; name = Cache_misses }
+    | str -> t_of_sexp (Sexp.of_string str)
+  ;;
+
+  let arg_type = Command.Arg_type.create of_string
+
+  let to_perf_event_config { config; name } =
```

Comment:
Stuff in core should be agnostic to backend, so we shouldn't have `perf`-specific bits here. These functions should live in `src/` (and they shouldn't rely on `to_string` from `core` files matching up to what `perf` expects).